homeassistant/components/sensor/metlink.py

```diff
@@ -0,0 +1,201 @@
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Get the Metlink public transport sensor."""
+    # journal contains [0] Station ID start, [1] Station ID destination
+    # [2] Station name start, and [3] Station name destination
+    stop_number = config.get('stop_number')
+    route_number = config.get('route_number')
+    add_devices([MetlinkSensor(stop_number, route_number)])
+
+
+def pretty_timestamp(t):
+    #
+    # python2's strptime doesn't handle %z so here's a workaround, from:
+    #   http://stackoverflow.com/questions/1101508/how-to-parse-dates-with-0400-timezone-string-in-python/23122493#23122493
+    #
+    ret = datetime.strptime(t[0:16], '%Y-%m-%dT%H:%M')
+    if t[18] == '+':
+        ret += timedelta(hours=int(t[19:22]), minutes=int(t[23:]))
+    elif t[18] == '-':
+        ret -= timedelta(hours=int(t[19:22]), minutes=int(t[23:]))
+
+    return ret.strftime("%I:%M%p %a")
+
+
+class MetlinkSensor(Entity):
+    """Implementation of an Metlink public transport sensor."""
+
+    def __init__(self, stop_number, route_number):
+        """Initialize the sensor."""
+        self.stop_number = stop_number
+        self.route_number = route_number
+        self._name = 'stop_{stop_number}_route_{route_number}'.format(
+            stop_number=stop_number,
+            route_number=route_number
+        )
+        self.update()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return "Route {}".format(self.route_number)
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        # return self.metlink_stop.services[0].departure_status
+        service = self.metlink_stop.next_service(
+            route_number=self.route_number)
+        if service and service.is_real_time:
+            return int(service.departure_seconds / 60)
+        elif service and service.display_departure:
+            return pretty_timestamp(service.display_departure)
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        attributes = {
+            'Stop': self.stop_number,
+            'Route': self.route_number,
+            'StopName': self.metlink_stop.stop_name,
+            'Latitude': self.metlink_stop.latitude,
+            'Longitude': self.metlink_stop.longitude
+        }
+
+        if self.next_service:
+            attributes.update(self.next_service.attributes())
+
+        return attributes
+
+    @property
+    def next_service(self):
+        return self.metlink_stop.next_service(route_number=self.route_number)
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return ICON
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit this state is expressed in."""
+        if self.next_service and self.next_service.is_real_time:
+            return "min"
+        else:
+            return "offline"
```

Comment:
I'm trying to reproduce the functionality of the road side signs.

When the bus' realtime eta is know, it shows "3 mins" etc.
When the bus realtime is not known, it instead shows the time from the schedule.

if I return available()=false, home-assistant doesn't show the `state` any more.